File: library/src/main/java/de/westnordost/osmfeatures/JsonUtils.java
package de.westnordost.osmfeatures;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

class JsonUtils {
    public interface Transformer<T> { T apply(Object item); }

    public static <T> List<T> parseList(JSONArray array, Transformer<T> t) throws JSONException
    {
        if(array == null) return new ArrayList<>(0);
        List<T> result = new ArrayList<>(array.length());
        for (int i = 0; i < array.length(); i++)
        {
            T item = t.apply(array.get(i));
            if(item != null) result.add(item);
        }
        return result;
    }

    public static List<String> parseCommaSeparatedList(String str, String remove)
    {
        if(str == null || str.isEmpty()) return new ArrayList<>(0);
        String[] array = str.split("\\s*,\\s*");
        List<String> result = new ArrayList<>(array.length);
        Collections.addAll(result, array);
        result.remove(remove);
        return result;
    }

    public static Map<String, String> parseStringMap(JSONObject map) throws JSONException
    {
        if(map == null) return new HashMap<>(1);
        Map<String, String> result = new HashMap<>(map.length());
        for (Iterator<String> it = map.keys(); it.hasNext(); )
        {
            String key = it.next().intern();
            result.put(key, map.getString(key));
        }
        return result;
    }

    // this is only necessary because Android uses some old version of org.json where
    // new JSONObject(new JSONTokener(inputStream)) is not defined...
    public static JSONObject createFromInputStream(InputStream inputStream) throws IOException
    {
        ByteArrayOutputStream result = new ByteArrayOutputStream();
        byte[] buffer = new byte[1024];
        int length;
        while ((length = inputStream.read(buffer)) != -1)
        {
            result.write(buffer, 0, length);
        }
        String jsonString = result.toString("UTF-8");
        return new JSONObject(jsonString);
    }
}

File: library/src/main/java/de/westnordost/osmfeatures/JsonUtils.kt
package de.westnordost.osmfeatures

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonPrimitive
import okio.Buffer
import okio.Source
import okio.buffer
internal object JsonUtils {

    @JvmStatic
    fun <T> parseList(array: JsonArray?, t: Transformer<T>): List<T> {
        return array?.mapNotNull { item -> t.apply(item) }.orEmpty()
    }

    @JvmStatic
    fun parseStringMap(map: JsonObject?): Map<String, String> {
        if (map == null) return HashMap(1)
        return map.map { (key, value) -> key.intern() to value.jsonPrimitive.toString()}.toMap().toMutableMap()
    }

    // this is only necessary because Android uses some old version of org.json where
    // new JSONObject(new JSONTokener(inputStream)) is not defined...
    @JvmStatic
    fun createFromSource(source: Source): JsonObject {
        val sink = Buffer()
        source.buffer().readAll(sink)

        return Json.decodeFromString<JsonObject>(sink.readUtf8())
    }

    fun interface Transformer<T> {
        fun apply(item: Any?): T
    }
}
